### PR TITLE
feat(CRT-150): client caching controller for clusters represented by KubeFedCluster CRs

### PR DESCRIPTION
**Pull request by MatousJobanek**
This PR brings a controller that watches `KubeFedCluster` CRs and caches clients (created based on the CRs) in a map. These cached clients can be used in both operators `host` and `member` for mutual communication.

Changes introduced in this PR: 
* a new controller that checks  `KubeFedCluster` CRs. The controller is implemented using `NewGenericInformerWithEventHandler` taken from KubeFed project
* a caching mechanism in `cluster.cache.go` that keeps clients of the KubeFed clusters in a map
* a service in `cluster.service.go` that takes care of the transition of `KubeFedCluster` data between the controller and the cache

### CODE REVIEW

File: pkg/cluster/service.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	return s.Log.
+		WithValues("Request.Namespace", cluster.Namespace, "Request.Name", cluster.Name)
+}
+
+func (s *KubeFedClusterService) buildClusterConfig(fedCluster *v1beta1.KubeFedCluster, fedNamespace string) (*rest.Config, error) {
+	clusterName := fedCluster.Name
+
+	apiEndpoint := fedCluster.Spec.APIEndpoint
+	if apiEndpoint == "" {
+		return nil, errors.Errorf("The api endpoint of cluster %s is empty", clusterName)
```
Comment (alexeykazakov) [inline pkg/cluster/service.go]:
```suggestion
		return nil, errors.Errorf("the api endpoint of cluster %s is empty", clusterName)
```

File: pkg/cluster/service.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	secretName := fedCluster.Spec.SecretRef.Name
+	if secretName == "" {
+		return nil, errors.Errorf("Cluster %s does not have a secret name", clusterName)
```
Comment (alexeykazakov) [inline pkg/cluster/service.go]:
```suggestion
		return nil, errors.Errorf("cluster %s does not have a secret name", clusterName)
```

File: pkg/cluster/service.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	secret := &v1.Secret{}
+	name := types.NamespacedName{Namespace: fedNamespace, Name: secretName}
+	err := s.Client.Get(context.TODO(), name, secret)
+	if err != nil {
+		return nil, err
+	}
+
+	token, tokenFound := secret.Data[util.TokenKey]
+	if !tokenFound || len(token) == 0 {
+		return nil, errors.Errorf("The secret for cluster %s is missing a non-empty value for %q", clusterName, util.TokenKey)
```
Comment (alexeykazakov) [inline pkg/cluster/service.go]:
```suggestion
		return nil, errors.Errorf("the secret for cluster %s is missing a non-empty value for %q", clusterName, util.TokenKey)
```

File: pkg/cluster/service_test.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+			Namespace: "test-namespace",
+			Labels:    labels,
+		},
+		Status: status,
+	}, secret
+}
+
+func TestAddKubeFedClusterAsMember(t *testing.T) {
+	// given
+	defer gock.OffAll()
```
Comment (alexeykazakov) [inline pkg/cluster/service_test.go]:
```suggestion
	defer gock.Off()
```
It's safer to use `gock.Off()` instead. Go runs tests in different packages in parallel by default. Killing all unmatched requests can cause race condition between tests from different packages and can cause test failures.

File: pkg/cluster/service_test.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+				assert.Equal(t, labels["namespace"], fedCluster.OperatorNamespace)
+			}
+			assert.Equal(t, status, *fedCluster.ClusterStatus)
+		})
+	}
+}
+
+func TestAddKubeFedClusterAsHost(t *testing.T) {
+	// given
+	defer gock.OffAll()
```
Comment (alexeykazakov) [inline pkg/cluster/service_test.go]:
```suggestion
	defer gock.Off()
```

File: pkg/cluster/service_test.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+				assert.Equal(t, labels["namespace"], fedCluster.OperatorNamespace)
+			}
+			assert.Equal(t, status, *fedCluster.ClusterStatus)
+		})
+	}
+}
+
+func TestAddKubeFedClusterFailsBecauseOfMissingSecret(t *testing.T) {
+	// given
+	defer gock.OffAll()
```
Comment (alexeykazakov) [inline pkg/cluster/service_test.go]:
```suggestion
	defer gock.All()
```

File: pkg/cluster/service_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+	// then
+	fedCluster, ok := cluster.GetFedCluster("east")
+	require.False(t, ok)
+	assert.Nil(t, fedCluster)
+}
+
+func TestAddKubeFedClusterFailsBecauseOfEmptySecret(t *testing.T) {
+	// given
+	defer gock.OffAll()
```
Comment (alexeykazakov) [inline pkg/cluster/service_test.go]:
```suggestion
	defer gock.Off()
```

File: pkg/cluster/service_test.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+	// then
+	fedCluster, ok := cluster.GetFedCluster("east")
+	require.False(t, ok)
+	assert.Nil(t, fedCluster)
+}
+
+func TestUpdateKubeFedCluster(t *testing.T) {
+	// given
+	defer gock.OffAll()
```
Comment (alexeykazakov) [inline pkg/cluster/service_test.go]:
```suggestion
	defer gock.Off()
```

File: pkg/cluster/service_test.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+	fedCluster, ok := cluster.GetFedCluster("east")
+	require.True(t, ok)
+	assert.Equal(t, cluster.Host, fedCluster.Type)
+	assert.Equal(t, "toolchain-host-operator", fedCluster.OperatorNamespace)
+	assert.Equal(t, statusFalse, *fedCluster.ClusterStatus)
+}
+
+func TestDeleteKubeFedCluster(t *testing.T) {
+	// given
+	defer gock.OffAll()
```
Comment (alexeykazakov) [inline pkg/cluster/service_test.go]:
```suggestion
	defer gock.Off()
```

File: pkg/controller/kubefedcluster_caching.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+func StartCachingController(mgr manager.Manager, stopChan <-chan struct{}) error {
+	cntrlName := "controller_kubefedcluster_with_cache"
+	clusterCacheService := cluster.KubeFedClusterService{
+		Client: mgr.GetClient(),
+		Log:    logf.Log.WithName(cntrlName),
+	}
+
+	_, clusterController, err := util.NewGenericInformerWithEventHandler(
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/kubefedcluster_caching.go]: After disusing it with Matous it turned out that K8s already takes care of it. When controller starts K8s starts a reconcile cycle for all the primary resources. So, we should be fine here.

File: pkg/cluster/service.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+// it's used for adding/updating/deleting
+type KubeFedClusterService struct {
+	Client client.Client
+	Log    logr.Logger
+}
+
+// AddKubeFedCluster takes the KubeFedCluster CR object,
+// creates FedCluster with a kube client and stores it in a cache
+func (s *KubeFedClusterService) AddKubeFedCluster(obj interface{}) {
+	cluster, ok := obj.(*v1beta1.KubeFedCluster)
```
Comment (xcoulon) [inline pkg/cluster/service.go]: these first lines are repeated in every "exposed" methods of this service. Maybe you could factorize them?

File: pkg/cluster/service.go [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+	secretName := fedCluster.Spec.SecretRef.Name
+	if secretName == "" {
+		return nil, errors.Errorf("Cluster %s does not have a secret name", clusterName)
+	}
+	secret := &v1.Secret{}
+	name := types.NamespacedName{Namespace: fedNamespace, Name: secretName}
+	err := s.Client.Get(context.TODO(), name, secret)
+	if err != nil {
+		return nil, err
```
Comment (xcoulon) [inline pkg/cluster/service.go]: do you need to wrap the error with a contextual message?

File: pkg/cluster/cache_whitebox_test.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+	// given
+	defer resetClusterCache()
+	fedCluster := newTestFedCluster("testCluster", v1.ConditionTrue)
+
+	// when
+	clusterCache.addFedCluster(fedCluster)
+
+	// then
+	assert.Len(t, clusterCache.clusters, 1)
+	assert.Equal(t, *fedCluster, *(clusterCache.clusters["testCluster"]))
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cluster/cache_whitebox_test.go]: actually, I can do it... I wanted to verify the actual values, but it's right that testing the pointers directly does the same :-) 0f895fb

File: pkg/cluster/cache_whitebox_test.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func TestGetCluster(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cluster/cache_whitebox_test.go]: good idea 6335143

File: pkg/cluster/cache_whitebox_test.go [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+	// when
+	returnedFedCluster, ok := clusterCache.getFedCluster("testCluster")
+
+	// then
+	assert.True(t, ok)
+	assert.Equal(t, *fedCluster, *returnedFedCluster)
+}
+
+func TestExportedGetCluster(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cluster/cache_whitebox_test.go]: well, they are different methods - that's why I created different tests... But I can refactor it slightly - is this better: b192c13 ?

File: pkg/cluster/service.go (hunk 11 above)
Comment (MatousJobanek) [reply pkg/cluster/service.go]: I was thinking about it when I was writing it, but I would safe max one line per method comparing to 7 new lines of the new function. That's why I decided to not do that...  But I see that it doesn't look good, so I changed it: 9147ad1

File: pkg/cluster/service.go (hunk 12 above)
Comment (MatousJobanek) [reply pkg/cluster/service.go]: done 717dcc1

**Pull request merged**